modules/svg/lottie_texture.cpp

```diff
@@ -0,0 +1,376 @@
+
+#include "lottie_texture.h"
+
+#include "core/os/memory.h"
+#include "core/variant/variant.h"
+
+#include <thorvg.h>
+
+void LottieTexture2D::_load_lottie_json() {
+	if (json.is_null()) {
+		return;
+	}
+	String lottie_str = json->get_parsed_text();
+	if (lottie_str.is_empty()) {
+		// don't sort keys, otherwise ThorVG can't load it
+		lottie_str = JSON::stringify(json->get_data(), "", false, true);
+	}
+	tvg::Result result = picture->load(lottie_str.utf8(), lottie_str.utf8().size(), "lottie", true);
+	if (result != tvg::Result::Success) {
+		ERR_FAIL_MSG(vformat("LottieTexture2D: Couldn't load Lottie: %s.",
+				result == tvg::Result::InvalidArguments				   ? "InvalidArguments"
+						: result == tvg::Result::NonSupport			   ? "NonSupport"
+						: result == tvg::Result::InsufficientCondition ? "InsufficientCondition"
+																	   : "Unknown Error"));
+	}
+}
+
+void LottieTexture2D::_update_image() {
+	if (origin_width < 0 && origin_height < 0) {
+		float fw, fh;
+		picture->size(&fw, &fh);
+		origin_width = fw;
+		origin_height = fh;
+	}
+	if (json.is_null() || frame_count <= 0) {
+		return;
+	}
+
+	int _rows = rows < 0 ? Math::ceil(Math::sqrt((float)frame_count)) : rows;
+	int _columns = Math::ceil(((float)frame_count) / _rows);
+
+	uint32_t w = MAX(1, round(origin_width * scale));
+	uint32_t h = MAX(1, round(origin_height * scale));
+
+	const uint32_t max_dimension = 16384;
+	if (w * _columns > max_dimension || h * _rows > max_dimension) {
+		WARN_PRINT(vformat(
+				String::utf8("LottieTexture2D: Target canvas dimensions %d×%d (with scale %.2f, rows %d, columns %d) exceed the max supported dimensions %d×%d. The target canvas will be scaled down."),
+				w, h, scale, _rows, _columns, max_dimension, max_dimension));
+		w = MIN(w, max_dimension / _columns);
+		h = MIN(h, max_dimension / _rows);
+		scale = MIN(w / origin_width, h / origin_height);
+	}
+	picture->size(w, h);
+
+	Ref<Image> image = Image::create_empty(w * _columns, h * _rows, false, Image::FORMAT_RGBA8);
+	uint32_t *buffer = (uint32_t *)memalloc(sizeof(uint32_t) * w * h);
+
+	tvg::Result res = sw_canvas->target(buffer, w, w, h, tvg::SwCanvas::ARGB8888S);
+	if (res != tvg::Result::Success) {
+		memfree(buffer);
+		ERR_FAIL_MSG("LottieTexture2D: Couldn't set target on ThorVG canvas.");
+	}
+	res = sw_canvas->push(tvg::cast(picture));
+	if (res != tvg::Result::Success) {
+		memfree(buffer);
+		ERR_FAIL_MSG("LottieTexture2D: Couldn't insert ThorVG picture on canvas.");
+	}
+
+	for (int row = 0; row < _rows; row++) {
+		for (int column = 0; column < _columns; column++) {
+			if (row * _columns + column >= frame_count) {
+				break;
+			}
+			float progress = ((float)(row * _columns + column)) / frame_count;
+			float current_frame = frame_begin + (frame_end - frame_begin) * progress;
+
+			res = animation->frame(current_frame);
+			if (progress == 0 || res == tvg::Result::Success) {
+				sw_canvas->update(picture);
+			}
+
+			res = sw_canvas->draw();
```

Comment:
I think the draw() call should also be followed by an update() call (inside lines 108 to 110), since rendering without updates wouldn't be expected.
```
			res = sw_canvas->draw();
			if (res != tvg::Result::Success) {
				memfree(buffer);
				ERR_FAIL_MSG("LottieTexture2D: Couldn't draw ThorVG pictures on canvas.");
			}
```